vllm/model_executor/models/qwen.py

```diff
@@ -26,11 +41,476 @@
+    emb_w = get_1d_sincos_pos_embed_from_grid(embed_dim // 2,
+                                              grid[1])  # (H*W, D/2)
+
+    emb = np.concatenate([emb_h, emb_w], axis=1)  # (H*W, D)
+    return emb
+
+
+def get_1d_sincos_pos_embed_from_grid(embed_dim, pos):
+    """
+    embed_dim: output dimension for each position
+    pos: a list of positions to be encoded: size (M,)
+    out: (M, D)
+    """
+    assert embed_dim % 2 == 0
+    omega = np.arange(embed_dim // 2, dtype=np.float32)
+    omega /= embed_dim / 2.
+    omega = 1. / 10000**omega  # (D/2,)
+
+    pos = pos.reshape(-1)  # (M,)
+    out = np.einsum('m,d->md', pos, omega)  # (M, D/2), outer product
+
+    emb_sin = np.sin(out)  # (M, D/2)
+    emb_cos = np.cos(out)  # (M, D/2)
+
+    emb = np.concatenate([emb_sin, emb_cos], axis=1)  # (M, D)
+    return emb
+
+
+class Resampler(nn.Module):
+    """
+    A 2D perceiver-resampler network with one cross attention layers by
+        (grid_size**2) learnable queries and 2d sincos pos_emb
+    Outputs:
+        A tensor with the shape of (grid_size**2, embed_dim)
+    """
+
+    def __init__(self,
+                 grid_size,
+                 embed_dim,
+                 num_heads,
+                 device,
+                 dtype,
+                 kv_dim=None,
+                 norm_layer=nn.LayerNorm):
+        super().__init__()
+        self.num_queries = grid_size**2
+        self.embed_dim = embed_dim
+        self.num_heads = num_heads
+        # NOTE - we need to directly initialize the device / dtype since we
+        # init this parameter out of a numpy array, so it defaults to CPU
+        # even though the model itself is initialized in a torch device context
+        # manager by default.
+        self.pos_embed = nn.Parameter(
+            torch.from_numpy(get_2d_sincos_pos_embed(embed_dim, grid_size)).to(
+                dtype=dtype, device=device).requires_grad_(False))
+
+        self.query = nn.Parameter(
+            torch.zeros(self.num_queries, embed_dim).to(device))
+        trunc_normal_(self.query, std=.02)
+
+        if kv_dim is not None and kv_dim != embed_dim:
+            self.kv_proj = nn.Linear(kv_dim, embed_dim, bias=False)
+        else:
+            self.kv_proj = nn.Identity()
+
+        self.attn = nn.MultiheadAttention(embed_dim, num_heads)
+        self.ln_q = norm_layer(embed_dim)
+        self.ln_kv = norm_layer(embed_dim)
+
+    def forward(self, x, attn_mask=None):
+        pos_embed = get_abs_pos(self.pos_embed, x.size(1))
+
+        x = self.kv_proj(x)
+        x = self.ln_kv(x).permute(1, 0, 2)
+
+        N = x.shape[1]
+        q = self.ln_q(self.query)
+        out = self.attn(self._repeat(q, N) + self.pos_embed.unsqueeze(1),
+                        x + pos_embed.unsqueeze(1),
+                        x,
+                        attn_mask=attn_mask)[0]
+        return out.permute(1, 0, 2)
+
+    def _repeat(self, query, N: int):
+        return query.unsqueeze(1).repeat(1, N, 1)
```

Comment:
Same for this one. You can check whether [Qwen-VL's implementation](https://huggingface.co/Qwen/Qwen-VL/blob/main/visual.py) is the same as [MiniCPM-V's](https://huggingface.co/openbmb/MiniCPM-Llama3-V-2_5/blob/main/resampler.py).

Reply:
Thanks for the pointer! There are small differences, mostly in caching and some extra stuff like post layer normalization and a final projection, but the version in 2.0 looks similar - I'll pull the `BaseResampler` out of the VLLM MiniCPM and make the last layer norm/projection optional so that we can use it here too

Reply:
Base Resampler / MiniCPM-V 2.0's resampler have been pulled out to a new file in the model executor layers [here](https://github.com/vllm-project/vllm/blob/4f25926c5ad103511d8a7ba49339bd5addb134ba/vllm/model_executor/layers/resampler.py)